--- a/chrome/browser/ui/BUILD.gn
+++ b/chrome/browser/ui/BUILD.gn
@@ -4279,8 +4279,8 @@ static_library("ui") {
       "views/tabs/browser_tab_strip_controller.h",
       "views/tabs/color_picker_view.cc",
       "views/tabs/color_picker_view.h",
-      "views/tabs/compound_tab_container.cc",
-      "views/tabs/compound_tab_container.h",
+      #"views/tabs/compound_tab_container.cc",
+      #"views/tabs/compound_tab_container.h",
       "views/tabs/dragging/drag_session_data.cc",
       "views/tabs/dragging/drag_session_data.h",
       "views/tabs/dragging/dragging_tabs_session.cc",
--- a/chrome/browser/ui/views/tabs/tab_strip.cc
+++ b/chrome/browser/ui/views/tabs/tab_strip.cc
@@ -59,7 +59,7 @@
 #include "chrome/browser/ui/view_ids.h"
 #include "chrome/browser/ui/views/frame/browser_view.h"
 #include "chrome/browser/ui/views/tabs/browser_tab_strip_controller.h"
-#include "chrome/browser/ui/views/tabs/compound_tab_container.h"
+//#include "chrome/browser/ui/views/tabs/compound_tab_container.h"
 #include "chrome/browser/ui/views/tabs/dragging/tab_drag_controller.h"
 #include "chrome/browser/ui/views/tabs/tab.h"
 #include "chrome/browser/ui/views/tabs/tab_container_impl.h"
@@ -135,10 +135,13 @@ std::unique_ptr<TabContainer> MakeTabCon
     TabStrip* tab_strip,
     TabHoverCardController* hover_card_controller,
     TabDragContext* drag_context) {
+  /*
   if (base::FeatureList::IsEnabled(tabs::kSplitTabStrip)) {
+
     return std::make_unique<CompoundTabContainer>(
         *tab_strip, hover_card_controller, drag_context, *tab_strip, tab_strip);
   }
+  */
   return std::make_unique<TabContainerImpl>(
       *tab_strip, hover_card_controller, drag_context, *tab_strip, tab_strip);
 }
