--- a/chrome/browser/ui/views/frame/browser_view.cc
+++ b/chrome/browser/ui/views/frame/browser_view.cc
@@ -5182,10 +5182,15 @@ void BrowserView::AddedToWidget() {
 
   if (!base::CommandLine::ForCurrentProcess()->HasSwitch("remove-tabsearch-button"))
   if (GetIsNormalType()) {
-    if (features::HasTabSearchToolbarButton()) {
+    if (features::HasTabSearchToolbarButton() && !tabs::AreHeliumVerticalTabsEnabled()) {
       tab_search_bubble_host_ = std::make_unique<TabSearchBubbleHost>(
           toolbar_->tab_search_button(), browser_.get(),
           tabstrip()->AsWeakPtr());
+    } else if (tabs::AreHeliumVerticalTabsEnabled()) {
+      tab_search_bubble_host_ = std::make_unique<TabSearchBubbleHost>(
+          BrowserElementsViews::From(browser_.get())
+              ->GetViewAs<TabStripControlButton>(kTabSearchButtonElementId),
+          browser_.get(), tabstrip()->AsWeakPtr());
     } else {
       tab_search_bubble_host_ = std::make_unique<TabSearchBubbleHost>(
           BrowserElementsViews::From(browser_.get())
